src/index.ts

```diff
@@ -81,22 +86,139 @@ export class Client {
+        return {
+          response: null,
+          body: {
+            errors: [
+              {
+                code: 'PAYMENT_METHOD_ERROR',
+                message: 'Error creating payment method',
+                field: 'paymentMethod'
+              }
+            ]
+          }
+        };
+      }
+
+      // Add the paymentMethod to the data
+      appendExtraData(data, 'paymentMethod', payload.paymentMethod.id);
+
+      // Send a request to Formspree server to handle the payment method
+      const response = await fetchImpl(url, {
+        ...request,
+        body: data
+      });
+      const responseData = await response.json();
+
+      // Handle server side errors
+      if (responseData.error) {
+        return {
+          response,
+          body: {
+            errors: responseData.errors
+          }
+        };
+      }
+
+      // Handle success response
+      if (responseData.next && responseData.ok) {
+        return {
+          response,
+          body: responseData
+        };
+      }
+
+      // Handle SCA
+      if (
+        responseData &&
+        responseData.stripe &&
+        responseData.stripe.requiresAction &&
+        responseData.resubmitKey
+      ) {
+        const stripeResult = await this.stripePromise.handleCardAction(
+          responseData.stripe.paymentIntentClientSecret
+        );
+
+        // Handle Stripe error
+        if (stripeResult.error) {
+          return {
+            response,
+            body: {
+              errors: [
+                {
+                  code: 'STRIPE_SCA_ERROR',
+                  message: 'Stripe SCA error',
+                  field: 'paymentMethod'
+                }
+              ]
+            }
+          };
+        } else {
+          if (!payload.paymentMethod.id) {
+            appendExtraData(data, 'paymentMethod', payload.paymentMethod.id);
+          }
+          appendExtraData(data, 'paymentIntent', stripeResult.paymentIntent.id);
+          appendExtraData(data, 'resubmitKey', responseData.resubmitKey);
+
+          // Resubmit the form with the paymentIntent and resubmitKey
+          const resSubmitResponse = await fetchImpl(url, {
+            ...request,
+            body: JSON.stringify({
+              paymentIntent: stripeResult.paymentIntent.id,
+              resubmitKey: responseData.resubmitKey
+            })
+          });
+          const resSubmitData = await resSubmitResponse.json();
+
+          // Handle success for resubmission
+          if (resSubmitData.next && resSubmitData.ok) {
+            return {
+              response: resSubmitResponse,
+              body: resSubmitData
+            };
+          }
+
+          // Handle server side errors for resubmission
+          if (resSubmitData.errors) {
+            return {
+              response: resSubmitResponse,
+              body: {
+                errors: resSubmitData.errors
+              }
+            };
+          }
```

Comment:
Rather than calling `appendExtraData`, then `fetchImpl` and finally handling errors / success (again) could we just call `submitForm()` again here, passing in the extra data to the `data` argument, and removing the `createPaymentMethod` opt?

Reply:
I try to avoid recursive functions as much as I can 😆